### ISSUE TITLE
[clang++] Clang not emitting diagnostic warning for possible integer overflow

### ISSUE THREAD

**Issue by lamarrr** [llvm/llvm-project#87941]
Consider the following code:

```cpp
#include "stdint.h"

void ST_WriteReg(uint16_t reg, uint8_t* bytes, uint32_t size);

void ST_SetCursor(uint16_t x, uint16_t y) {
    uint8_t param[4];
    param[0] = 0x00;
    param[1] = 0x00 + x;
    param[2] = 0x00;
    param[3] = 0xEF + x;
    ST_WriteReg(0, param, 4);
    param[0] = 0x00;
    param[1] = 0x00 + y;
    param[2] = 0x00;
    param[3] = 0xEF + y;
    ST_WriteReg(0, param, 4);
}
```

even with all warnings (-Wall -Wextra -Wpedantic) turned on, no warning is emitted for assigning the promoted (u16) to the u8 array.
is this the intended behavior? if not, I'd like to take on this.

**Comment by firewave** [comment]
Try `-Weverything` for *all* warnings to be enabled:
```
<source>:8:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
    8 |     param[1] = 0x00 + x;
      |              ~ ~~~~~^~~
<source>:10:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
   10 |     param[3] = 0xEF + x;
      |              ~ ~~~~~^~~
<source>:13:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
   13 |     param[1] = 0x00 + y;
      |              ~ ~~~~~^~~
<source>:15:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
   15 |     param[3] = 0xEF + y;
      |              ~ ~~~~~^~~
```
https://godbolt.org/z/jYPe4fvjj

**Comment by lamarrr** [comment]
> Try `-Weverything` for _all_ warnings to be enabled:
> 
> ```
> <source>:8:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
>     8 |     param[1] = 0x00 + x;
>       |              ~ ~~~~~^~~
> <source>:10:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
>    10 |     param[3] = 0xEF + x;
>       |              ~ ~~~~~^~~
> <source>:13:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
>    13 |     param[1] = 0x00 + y;
>       |              ~ ~~~~~^~~
> <source>:15:21: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
>    15 |     param[3] = 0xEF + y;
>       |              ~ ~~~~~^~~
> ```
> 
> https://godbolt.org/z/jYPe4fvjj

thanks!
